cpp/FileTransferAgent.cpp

```diff
@@ -250,3 +277,101 @@ void Snowflake::Client::FileTransferAgent::compressSourceFile(
+{
+  m_executionResults = new FileTransferExecutionResult(CommandType::DOWNLOAD,
+    m_largeFilesMeta.size() + m_smallFilesMeta.size());
+
+  std::string createDirIfNotExist = "mkdir -p " +
+    std::string(response.GetLocalLocation());
+  const int err = system(createDirIfNotExist.c_str());
+  if (err != 0)
+  {
+    throw;
+  }
+
+  if (m_largeFilesMeta.size() > 0)
+  {
+    for (size_t i=0; i<m_largeFilesMeta.size(); i++)
+    {
+      m_executionResults->SetFileMetadata(&m_largeFilesMeta[i], i);
+      RemoteStorageRequestOutcome outcome = downloadSingleFile(m_storageClient,
+                                                               &m_largeFilesMeta[i], i);
+
+      if (outcome == RemoteStorageRequestOutcome::TOKEN_EXPIRED)
+      {
+        renewToken(command);
+        i--;
+      }
+    }
+  }
+
+  if (m_smallFilesMeta.size() > 0)
+  {
+    downloadFilesInParallel(command);
+  }
+}
+
+void Snowflake::Client::FileTransferAgent::downloadFilesInParallel(std::string *command)
+{
+  SF_MUTEX_HANDLE tokenRenewMutex;
+  _mutex_init(&tokenRenewMutex);
+
+  Snowflake::Client::Util::ThreadPool tp((unsigned int)response.getParallel());
+  for (size_t i=0; i<m_smallFilesMeta.size(); i++)
+  {
+    unsigned int resultIndex = i + m_largeFilesMeta.size();
+    FileMetadata * metadata = &m_smallFilesMeta[i];
+    m_executionResults->SetFileMetadata(&m_smallFilesMeta[i], resultIndex);
+    tp.AddJob([metadata, resultIndex, &tokenRenewMutex, command, this]()->void {
+        do
+        {
+          RemoteStorageRequestOutcome outcome = downloadSingleFile(m_storageClient, metadata,
+                                                                 resultIndex);
+          if (outcome == RemoteStorageRequestOutcome::TOKEN_EXPIRED)
+          {
+            _mutex_lock(&tokenRenewMutex);
+            this->renewToken(command);
+            _mutex_unlock(&tokenRenewMutex);
+          }
+          else
+          {
+            break;
+          }
+        } while (true);
+    });
+  }
+
+  // wait till all jobs have been finished
+  tp.WaitAll();
+
+  _mutex_term(&tokenRenewMutex);
+}
+
+RemoteStorageRequestOutcome Snowflake::Client::FileTransferAgent::downloadSingleFile(
+  IStorageClient *client,
+  FileMetadata *fileMetadata,
+  unsigned int resultIndex)
+{
+  EncryptionProvider::decryptFileKey(fileMetadata, response.getEncryptionMaterial());
+  std::string dstNameFull = std::string(response.GetLocalLocation()) + "/" +
+    fileMetadata->destFileName;
+
+  std::basic_fstream<char> dstFile(dstNameFull.c_str(),
+                                std::ios_base::out | std::ios_base::binary);
+
+  Crypto::CipherIOStream decryptOutputStream(
+                               dstFile,
+                               Crypto::CryptoOperation::DECRYPT,
+                               fileMetadata->encryptionMetadata.fileKey,
+                               fileMetadata->encryptionMetadata.iv, 128);
```

Comment:
128 can be hardcoded here? There is a case of 256 encryption.

Reply:
It is random value. I should not use 128 here, it is confusing. This value is encryption block size, i.e. for one encryption call, how many bytes of data is encrypted. I will change this to a MACRO ENC_DEC_BLOCK_SIZE. I think hard code value is fine. I did not see a point to add a parameter just for that.

Reply:
ok, I just wondered it is related to the parameter CLIENT_ENCRYPTION_KEY_SIZE, which can be 128 or 256.
Since 128 is used in multiple places, please define a macro for better doc. It doesn't need to be a parameter.